File: server/pkg/featureflag/static_provider.go
package featureflag

import (
	"context"
	"slices"
	"sync"
)

type Rule struct {
	Default bool

	Variant string

	Allow []string

	AllowBy string

	Deny []string

	DenyBy string

	Percent *PercentRollout
}

type PercentRollout struct {
	Percent int

	By string
}

type StaticProvider struct {
	mu    sync.RWMutex
	rules map[string]Rule
}

func NewStaticProvider() *StaticProvider {
	return &StaticProvider{rules: map[string]Rule{}}
}

func (*StaticProvider) Name() string { return "static" }

func (p *StaticProvider) Set(key string, rule Rule) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.rules[key] = rule
}

func (p *StaticProvider) LoadRules(rules map[string]Rule) {
	clone := make(map[string]Rule, len(rules))
	for k, v := range rules {
		clone[k] = v
	}
	p.mu.Lock()
	defer p.mu.Unlock()
	p.rules = clone
}

func (p *StaticProvider) Keys() []string {
	p.mu.RLock()
	defer p.mu.RUnlock()
	out := make([]string, 0, len(p.rules))
	for k := range p.rules {
		out = append(out, k)
	}
	slices.Sort(out)
	return out
}

func (p *StaticProvider) Lookup(ctx context.Context, key string) (Decision, bool) {
	p.mu.RLock()
	rule, ok := p.rules[key]
	p.mu.RUnlock()
	if !ok {
		return Decision{}, false
	}
	ec := EvalContextFrom(ctx)
	return evaluateRule(key, rule, ec), true
}

func evaluateRule(key string, rule Rule, ec EvalContext) Decision {

	denyBy := orDefault(rule.DenyBy, "user_id")
	if len(rule.Deny) > 0 {
		if v, ok := ec.Lookup(denyBy); ok && slices.Contains(rule.Deny, v) {
			return decisionFromRule(key, rule, false, ReasonStatic)
		}
	}

	allowBy := orDefault(rule.AllowBy, "user_id")
	if len(rule.Allow) > 0 {
		if v, ok := ec.Lookup(allowBy); ok && slices.Contains(rule.Allow, v) {
			return decisionFromRule(key, rule, true, ReasonStatic)
		}
	}

	if rule.Percent != nil {
		by := orDefault(rule.Percent.By, "user_id")
		identifier, _ := ec.Lookup(by)

		if inPercent(key, identifier, rule.Percent.Percent) {
			return decisionFromRule(key, rule, true, ReasonPercent)
		}
		return decisionFromRule(key, rule, false, ReasonPercent)
	}

	return decisionFromRule(key, rule, rule.Default, ReasonStatic)
}

func decisionFromRule(key string, rule Rule, enabled bool, reason Reason) Decision {

	variant := boolToVariant(enabled)
	if enabled && rule.Variant != "" {
		variant = rule.Variant
	}
	return Decision{
		Key:     key,
		Enabled: enabled,
		Variant: variant,
		Reason:  reason,
		Source:  "static",
	}
}

func orDefault(v, def string) string {
	if v == "" {
		return def
	}
	return v
}

File: packages/core/feature-flags/static-provider.ts
import type { Decision, EvalContext, Provider, Rule } from './types';
import { inPercent } from './hash';

export class StaticProvider implements Provider {
  readonly name = 'static';
  private rules: Map<string, Rule>;

  constructor(rules: Readonly<Record<string, Rule>> = {}) {
    this.rules = new Map(Object.entries(rules));
  }

  set(key: string, rule: Rule): void {
    this.rules.set(key, rule);
  }

  loadRules(rules: Readonly<Record<string, Rule>>): void {
    this.rules = new Map(Object.entries(rules));
  }

  keys(): string[] {
    return Array.from(this.rules.keys()).sort();
  }

  lookup(key: string, ctx: EvalContext): Decision | undefined {
    const rule = this.rules.get(key);
    if (!rule) return undefined;
    return evaluateRule(key, rule, ctx);
  }
}

function evaluateRule(key: string, rule: Rule, ctx: EvalContext): Decision {
  const denyBy = rule.denyBy ?? 'user_id';
  if (rule.deny && rule.deny.length > 0) {
    const v = lookupAttr(ctx, denyBy);
    if (v && rule.deny.includes(v)) {
      return decisionFromRule(key, rule, false, 'static');
    }
  }

  const allowBy = rule.allowBy ?? 'user_id';
  if (rule.allow && rule.allow.length > 0) {
    const v = lookupAttr(ctx, allowBy);
    if (v && rule.allow.includes(v)) {
      return decisionFromRule(key, rule, true, 'static');
    }
  }

  if (rule.percent) {
    const by = rule.percent.by ?? 'user_id';
    const ident = lookupAttr(ctx, by) ?? '';
    const enabled = inPercent(key, ident, rule.percent.percent);
    return decisionFromRule(key, rule, enabled, 'percent');
  }

  return decisionFromRule(key, rule, rule.default ?? false, 'static');
}

function decisionFromRule(
  key: string,
  rule: Rule,
  enabled: boolean,
  reason: Decision['reason'],
): Decision {
  let variant = boolToVariant(enabled);
  if (enabled && rule.variant && rule.variant.length > 0) {
    variant = rule.variant;
  }
  return {
    key,
    enabled,
    variant,
    reason,
    source: 'static',
  };
}

function boolToVariant(b: boolean): string {
  return b ? 'on' : 'off';
}

function lookupAttr(ctx: EvalContext, name: string): string | undefined {
  if (name === 'user_id') return nonEmpty(ctx.userId);
  if (name === 'workspace_id') return nonEmpty(ctx.workspaceId);
  return nonEmpty(ctx.attributes?.[name]);
}

function nonEmpty(v: string | undefined): string | undefined {
  return v && v.length > 0 ? v : undefined;
}
